Create base class for serialization tests

mziccard opened the pull request:
This fixes #657 and #676 by adding a base class for serialization tests as well as a `SerializationTest` to `gcloud-java-core`.

To achieve this I had to add a new execution goal to our build (`text-jar`) and add core's test jar as a test dependency to all our modules' poms.

ajkannan commented on gcloud-java-resourcemanager/src/test/java/com/google/gcloud/resourcemanager/SerializationTest.java:
```diff
@@ -16,26 +16,17 @@
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
 import com.google.gcloud.Identity;
+import com.google.gcloud.BaseSerializationTest;
```
Nit: this should be moved up above the Identity import

ajkannan commented on gcloud-java-core/src/test/java/com/google/gcloud/BaseSerializationTest.java:
```diff
@@ -0,0 +1,64 @@
+
+  @SuppressWarnings("unchecked")
+  public <T> T serializeAndDeserialize(T obj)
+      throws IOException, ClassNotFoundException {
```
Nit: fits on previous line

ajkannan commented on gcloud-java-core/src/test/java/com/google/gcloud/SerializationTest.java:
```diff
@@ -0,0 +1,33 @@
+
+  private static final PageImpl<String> PAGE =
+      new PageImpl<>(null, "cursor", ImmutableList.of("string1", "string2"));
+  private static final RetryParams RETRY_PARAMS = RetryParams.defaultInstance();
```
We can add this in a separate PR or here (your choice), but `Identity` (in core, for IAM) implements Serializable and can be added here.

ajkannan replied:
Could we also try serializing an `ExceptionHandler` here?

ajkannan replied:
And we could also add in the AuthCredentials' restorable states.  (we can create an issue and do this later if you prefer not to add it in this PR).

ajkannan commented on gcloud-java-storage/src/test/java/com/google/gcloud/storage/SerializationTest.java:
```diff
@@ -77,37 +73,21 @@
-    options = options.toBuilder()
+    StorageOptions otherOptions = options.toBuilder()
         .projectId("p2")
         .retryParams(RetryParams.defaultInstance())
```
Unrelated to this PR: the `.retryParams(RetryParams.defaultInstance())` lines (present in a couple of the modules' serialization tests) doesn't test anything new anymore (compared to the `<Service>Options.defaultInstance()` options) because we switched retry params to be the default instance if none are specified (in ServiceOptions).  We can probably remove or switch those retry params to something else.

ajkannan commented:
Just one or two nits from me + a couple outdated things that are unrelated to your PR (can either be fixed here or in the future if we create an issue).

mziccard commented:
@ajkannan thanks for the feedback, you are right on everything so I tried to address all comments in 967c5c5

ajkannan commented on gcloud-java-core/src/test/java/com/google/gcloud/SerializationTest.java:
```diff
@@ -18,16 +18,60 @@
+
+  @Test
+  public void testAuthCredentialState() throws IOException, ClassNotFoundException {
+    AuthCredentials credentials = AuthCredentials.createApplicationDefaults();
```
This causes an error on Travis because we don't set GOOGLE_APPLICATION_CREDENTIALS or log into the Google Cloud SDK on Travis.

ajkannan commented on gcloud-java-core/src/test/java/com/google/gcloud/BaseSerializationTest.java:
```diff
@@ -0,0 +1,79 @@
+  /**
+   * Returns all objects for which correct serialization must be tested.
+   */
+  public abstract Serializable[] serializableObjects();
```
This could be protected, right?  Also, do you think we could have an additional `protected abstract` method called `restorableObjects` which returns an Array of restorableObjects?  Then the `assertRestorable` method could work like the `testSerializableObjects` method

mziccard replied:
> This could be protected, right?
Yes

> do you think we could have an additional protected abstract method called restorableObjects

No, not all modules have restorable objects. To implement the same behavior we should have a non-abstract `restorableObjects` that returns an empty array and a test that passes without doing nothing, I don't like this idea.

aozarov replied:
> No, not all modules have restorable objects. 

True, and I don't think we should make such method non-abstract however I think overriding this method and returning an empty array (or null if we care) when "don't-care" is not a big deal.

And I think the benefits of making it obvious that one should also test the "serialization" of Restorables if they exists and have it done in a common way outweigh the disadvantage.

ajkannan replied:
I see your point @mziccard, but (in my opinion) the positives @aozarov mentioned outweigh the negative.

ajkannan commented on gcloud-java-core/src/main/java/com/google/gcloud/ExceptionHandler.java:
```diff
@@ -259,6 +260,27 @@ boolean shouldRetry(Exception ex) {
+        && Objects.equals(retriableExceptions, other.retriableExceptions)
+        && Objects.equals(nonRetriableExceptions, other.nonRetriableExceptions)
+        && Objects.equals(retryInfo, other.retryInfo);
+
```
Nit: extra line

ajkannan commented:
Nice, looks like Travis is passing now.  Just finished another pass with two more comments.

mziccard commented:
Guys thanks for the feedback, looks much nicer now. I think I addressed all comments but please have a look.

aozarov commented on gcloud-java-core/src/main/java/com/google/gcloud/AuthCredentials.java:
```diff
@@ -276,6 +276,16 @@ public GoogleCredentials credentials() {
       public AuthCredentials restore() {
         return INSTANCE;
       }
+
```
Shall we make NoAuthCredentials constructor private?

aozarov commented:
Nice, looks good to me but I will let @ajkannan merge it after looking more carefully.

aozarov commented on gcloud-java-core/src/test/java/com/google/gcloud/SerializationTest.java:
```diff
@@ -0,0 +1,74 @@
+
+  @Override
+  protected Serializable[] serializableObjects() {
+    return new Serializable[]{EXCEPTION_HANDLER, IDENTITY, PAGE, RETRY_PARAMS};
```
can you also add one for `IamPolicy` (by have something that extends it) and also for `BaseServiceException`?

aozarov commented:
I looked a deepr look. Looks great! Feel free to merge after fixing the 2 issues mentioned.

mziccard commented:
@aozarov I addressed the comments. I added `equals` and `hashCode` method to `BaseServiceException` and added the service exceptions to all serialization tests. Please take one more look before we merge this in.
